pkg/services/authn/grpcutils/grpc_authenticator.go

```diff
@@ -19,101 +19,79 @@ import (
 	"github.com/grafana/grafana/pkg/setting"
 )
 
-var once sync.Once
-
 func NewInProcGrpcAuthenticator() interceptors.Authenticator {
+	return newAuthenticator(
+		authn.NewDefaultAuthenticator(
+			authn.NewUnsafeAccessTokenVerifier(authn.VerifierConfig{}),
+			authn.NewUnsafeIDTokenVerifier(authn.VerifierConfig{}),
+		),
+		tracing.NewNoopTracerService(),
+	)
+}
+
+func NewAuthenticator(cfg *GrpcServerConfig, tracer tracing.Tracer) interceptors.Authenticator {
+	client := http.DefaultClient
+	if cfg.AllowInsecure {
+		client = &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}}
+	}
+
+	kr := authn.NewKeyRetriever(authn.KeyRetrieverConfig{
+		SigningKeysURL: cfg.SigningKeysURL,
+	}, authn.WithHTTPClientKeyRetrieverOpt(client))
+
 	auth := authn.NewDefaultAuthenticator(
-		authn.NewUnsafeAccessTokenVerifier(authn.VerifierConfig{}),
-		authn.NewUnsafeIDTokenVerifier(authn.VerifierConfig{}),
+		authn.NewAccessTokenVerifier(authn.VerifierConfig{AllowedAudiences: cfg.AllowedAudiences}, kr),
+		authn.NewIDTokenVerifier(authn.VerifierConfig{}, kr),
 	)
 
+	return newAuthenticator(auth, tracer)
+}
+
+func NewAuthenticatorWithFallback(cfg *setting.Cfg, reg prometheus.Registerer, tracer tracing.Tracer, fallback interceptors.Authenticator) interceptors.Authenticator {
+	return &authenticatorWithFallback{
+		authenticator: NewAuthenticator(ReadGrpcServerConfig(cfg), tracer),
+		fallback:      fallback,
+		tracer:        tracer,
+		metrics:       newMetrics(reg),
+	}
+}
+
+func newAuthenticator(auth authn.Authenticator, tracer tracing.Tracer) interceptors.Authenticator {
 	return interceptors.AuthenticatorFunc(func(ctx context.Context) (context.Context, error) {
+		ctx, span := tracer.Start(ctx, "grpcutils.Authenticate")
+		defer span.End()
+
 		md, ok := metadata.FromIncomingContext(ctx)
 		if !ok {
 			return nil, errors.New("missing metedata in context")
 		}
 
 		info, err := auth.Authenticate(ctx, authn.NewGRPCTokenProvider(md))
 		if err != nil {
+			span.RecordError(err)
 			return ctx, err
 		}
 
+		span.SetAttributes(attribute.String("subject", info.GetUID()))
```

Comment:
It's really convenient for debugging obo calls to have both the service ID as well.
Could you please add it?

Reply:
We don't have a way to get the service id from AuthInfo when calling with both. We need to add that to the interface and implementation, a way to guarantee that we get the service id. 

We discussed it a bit here but lets create and issue for it and add a todo here https://github.com/grafana/authlib/pull/113#issuecomment-2500046887